File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        int low = 0, high =0, m = s.length(), n= t.length() , res=Integer.MAX_VALUE, start=0, end=0;
        
        Map<Character, Integer> have = new HashMap<>();
        Map<Character, Integer> need = new HashMap<>();

        for(char ch : t.toCharArray()){
            need.put(ch, need.getOrDefault(ch, 0)+1);            
        }

        while(high< m ){
            have.put(s.charAt(high), have.getOrDefault(s.charAt(high), 0)+1);
            while(correctSubStr(have, need)){
                if(res > high-low+1){
                    res = high-low+1;
                    start = low;
                    end = high;
                }
                
                low++;
                have.put(s.charAt(low-1), have.get(s.charAt(low-1))-1);
                if(have.get(s.charAt(low-1))==0){
                    have.remove(s.charAt(low-1));
                }
            }
            high++;
        }

        
        if(res==Integer.MAX_VALUE){
            return "";
        }
        else{
            return s.substring(start, end+1);
        }
    }
    boolean correctSubStr(Map<Character, Integer> have, Map<Character, Integer> need){
        for(char c : need.keySet()){
            if(have.getOrDefault(c, 0) < need.get(c)){
                return false; 
            }
        }           
        return true;    //assume valid SubStr
    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        low = 0
        reqd = len(t)
        start = low
        res = float('inf')
        need = defaultdict(int) # maps char to their required occurences
#need tells us how many more char are required so when we add a char to our window simply decrement the count of that char and when we remove a char from our window increment the count of that char
        for i in t :
            need[i] += 1
        for high in range (0, len(s)):
            if need[s[high]] > 0:  # requirement of ch not currently fulfilled
                reqd -= 1
            need[s[high]] -= 1  #always do bcz it helps in window shrinking later
            while reqd == 0: # valid substr containing all reqd char found
                # as valid substr found so now we try to find the min valid substr
                if (high-low+1) < res: # store the len if curr valid substr len < old valid substr len
                    res = high -low+1
                    start = low
                #start window shrinking
                need[s[low]] +=1
                low+=1   # low char may or maynot be part of reqd char
                if need[s[low-1]] > 0: # s[low] is a reqd char
                    reqd+=1         # bcz we removed a reqd char
            high+=1
        if res == float('inf'):
            return ""
        return s[start: start+res]